### ISSUE TITLE
transport: add in memory transport

### ISSUE THREAD

**Issue by p-shahi** [libp2p/go-libp2p#2293]
Add an in memory transport (similar to rust-libp2p) which can be used for testing https://docs.rs/libp2p/latest/libp2p/core/transport/struct.MemoryTransport.html

**Comment by Wondertan** [comment]
Do I understand correctly that this should substitute mocknet ? 😃

**Comment by marten-seemann** [comment]
> Do I understand correctly that this should substitute mocknet ? 😃

Yes indeed. Mocknet shouldn't mock the swarm, only connections / streams.

**Comment by MarcoPolo** [comment]
We have a lot of open questions, so removing help wanted for now.

**Comment by sukunrt** [comment]
The requirement is ambiguous.

We need to decide whether we want an in memory transport for testing where we can fake IP address and possibly simulate NAT or we want to implement a `/memory` transport like the one implemented by rust-libp2p. 

`/memory` is useful for testing protocols which do not depend on specific remote or local addresses but it cannot be used for testing something like AutoNATv2 without changing the `/memory` multiaddress intention. @mxinden is my understanding correct? 

The effort required in both the cases seems similar.

**Comment by mxinden** [comment]
> `/memory` is useful for testing protocols which do not depend on specific remote or local addresses but it cannot be used for testing something like AutoNATv2 without changing the `/memory` multiaddress intention. @mxinden is my understanding correct?

That is correct. rust-libp2p's memory transport is very simple with a single uniform address space. It does not have any more complex mechanisms, e.g. NATs.

> The effort required in both the cases seems similar.

I would assume the opposite. While I think it is very simple to port rust-libp2p's `MemoryTransport` to go-libp2p, I expect building a Transport "where we can fake IP address and possibly simulate NAT" is much harder.

**Comment by sukunrt** [comment]
>I expect building a Transport "where we can fake IP address and possibly simulate NAT" is much harder.

I think you're right. I'll change my stance to 
Building a transport where we fake IP address is a little more work than adding a `/memory` transport
simulating NATs is a lot more work.

**Comment by MarcoPolo** [comment]
I think a in-memory transport without faking IP addresses or simulating NATs is still useful for testing protocols. 

I think simulating NATs should be a separate task.

I'm going to readd the help wanted label since I think we have a clearer idea of what we want here. Which is essentially the equivalent of rust-libp2p's `MemoryTransport`

**Comment by pyropy** [comment]
Could I claim this issue? Seems like a good way to get me started contributing to the go-libp2p

**Comment by pyropy** [comment]
Could I claim this issue? I think it would be a good intro for me to start contributing to go-libp2p

**Comment by MarcoPolo** [comment]
go for it

**Comment by pyropy** [comment]
Hey @MarcoPolo could you assign issue on me and maybe refer me to someone who could take a look at this 

https://github.com/multiformats/go-multiaddr/pull/256

**Comment by sukunrt** [comment]
This is mostly superseded by #3262

Does that PR work for you?

**Comment by PedrobyJoao** [comment]
Ahh, yeah, probably it also works..

For now, I just gave up of mocking (or having them in-memory) the connections as I was concerned if mocking the conns would make my tests any less reliable (it doesn't test business logic of other components but specifically of my lipb2p wrapper library)

anyway, #3262 would probably do the job also

thanks for the response btw

**Comment by sukunrt** [comment]
we don't need this as we have simnet now.
